azkaban-exec-server/src/main/java/azkaban/container/FlowContainer.java

```diff
@@ -805,12 +813,60 @@ public static void deleteSymlinkedFile(final Path symlinkedFilePath)
   }
 
   /**
-   * Uploads the log file to the db for persistence.
-   * @param execId execution id of the flow.
+   * setup logger and execution dir for the flowId
    */
-  private void uploadLogFile(final int execId) {
+  private void createLogger() {
+    Logger rootLogger = Logger.getRootLogger();
+
+    for (Object o: Collections.list(rootLogger.getAllAppenders())) {
+      if (o instanceof FileAppender) {
+        FileAppender appender = (FileAppender) o;
+        logFile = new File(appender.getFile());
+        break;
+      }
+    }
+
     try {
-      this.executionLogsLoader.uploadLogFile(execId, "", 0, FlowContainer.logFile);
+      if (this.logFile == null) {
+        logFile = new File(DEFAULT_LOG_FILE_NAME);
+        this.fileAppender = new FileAppender(DEFAULT_LAYOUT, logFile.getAbsolutePath(), true);
+        rootLogger.addAppender(this.fileAppender);
+        logger.info("Attached new File appender for flow");
+      }
+
+      if (this.azkabanProps.getBoolean(Constants.ConfigurationKeys.AZKABAN_LOGGING_KAFKA_ENABLED,
+          false)) {
+        // Keep the names consistent as what we did in uploadLogFile()
+        this.kafkaLog4jAppender =
+            KafkaLog4jUtils.getAzkabanFlowKafkaLog4jAppender(this.azkabanProps,
+            String.valueOf(this.execId), "");
+
+        if (this.kafkaLog4jAppender != null) {
+          rootLogger.addAppender(this.kafkaLog4jAppender);
+          rootLogger.setAdditivity(false);
+          logger.info("Attached new Kafka appender for flow");
+        }
+      }
+    } catch (final Exception e) {
+      logger.error("Could not open log file: " + logFile.getAbsolutePath(), e);
+    }
+  }
+
+  private static void removeAppender(final Appender appender) {
+    Logger rootLogger = Logger.getRootLogger();
+    if (appender != null) {
+      rootLogger.removeAppender(appender);
+      appender.close();
```

Comment:
do we need to handle exception here?

Reply:
Good catch! It should be handled well otherwise uploadLogFile will not be called if exception occurs.
I will address it in next follow-up PR tomorrow b/c FlowContainer&JobRunner&FlowRunner all have the removeAppender logic. It is out of the scope of this PR if touching JobRunner&FlowRunner.